src/libraries/System.Private.CoreLib/src/System/Globalization/IcuLocaleData.cs

```diff
@@ -4510,70 +3835,59 @@ internal static string GetConsoleUICulture(string cultureName)
+            for (int i = 0; i < name.Length; ++i) {
+                char ch = name[i];
+                if (ch > 'z')
+                    return -1;
 
-            name = CultureData.AnsiToLower(name);
+                lower_case[i] = (byte)(ch | 0x20);
+            }
 
-            // Binary search the array until we have only a couple of elements left and then
-            // just walk those elements.
-            while ((right - left) > 3)
-            {
-                index = ((right - left) / 2) + left;
+            ReadOnlySpan<byte> lname = lower_case;
 
-                Debug.Assert(index < s_localeNamesIndices.Length - 1);
-                result = CompareOrdinal(name, c_localeNames, s_localeNamesIndices[index], s_localeNamesIndices[index + 1] - s_localeNamesIndices[index]);
-                if (result == 0)
-                {
-                    return index;
-                }
-                else if (result < 0)
-                {
-                    right = index;
-                }
-                else
-                {
-                    left = index;
-                }
-            }
+            int lo = 0;
+            int hi = s_localesNamesIndexes.Length - 1;
 
-            // Walk the remaining elements (it'll be 3 or fewer).
-            for (; left <= right; left++)
-            {
-                Debug.Assert(left < s_localeNamesIndices.Length - 1);
-                if (CompareOrdinal(name, c_localeNames, s_localeNamesIndices[left], s_localeNamesIndices[left + 1] - s_localeNamesIndices[left]) == 0)
-                {
-                    return (left);
+            // Binary search the array
+            while (lo <= hi) {
+                int i = lo + ((hi - lo) >> 1);
+
+                int order = GetCultureName(i).SequenceCompareTo(lname);
+
+                if (order == 0) return i;
+                if (order < 0) {
+                    lo = i + 1;
+                } else {
+                    hi = i - 1;
                 }
             }
 
-            // couldn't find culture name
-            return -1;
+            return ~lo;
         }
 
-        // optimized to avoid parameters checking
-        private static int CompareOrdinal(string s1, string s2, int index, int length)
+        private static ReadOnlySpan<byte> GetCultureName(int localeNameIndice)
         {
-            int count = s1.Length;
-            if (count > length)
-                count = length;
-
-            int i = 0;
-            while (i < count && s1[i] == s2[index + i])
-                i++;
+            ushort index = s_localesNamesIndexes[localeNameIndice];
+            return CultureNames.Slice(index >> 4, index & 0xF);
+        }
 
-            if (i < count)
-                return (int)(s1[i] - s2[index + i]);
+        private static string GetString(ReadOnlySpan<byte> buffer)
+        {
+            string result = string.FastAllocateString(buffer.Length);
```

Comment:
Can it be just `return Encoding.ASCII.GetString(buffer);` ?

Reply:
No, it has very complicated dependencies and it's not used by anything else. We could move it to `ASCIIUtility` later if there is another use for it.